src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
+{
+    internal static class CodeActionHelpers
+    {
+        public static async Task<IEnumerable<CodeAction>> GetCodeActionsAsync(
+            Document document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
+                document, codeFixService,
+                codeRefactoringService, selection,
+                cancellationToken).ConfigureAwait(false);
+
+            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
+                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+
+            return codeActions;
+        }
+
+        public static async Task<(ImmutableArray<CodeFixCollection>, ImmutableArray<CodeRefactoring>)> GetCodeFixesAndRefactoringsAsync(
+            Document document,
+            ICodeFixService codeFixService,
+            ICodeRefactoringService codeRefactoringService,
+            LSP.Range selection,
+            CancellationToken cancellationToken)
+        {
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
+
+            var codeFixCollectionsTask = Task.Run(
+                () => codeFixService.GetFixesAsync(document, textSpan, includeSuppressionFixes: true, cancellationToken));
+            var codeRefactoringsTask = Task.Run(
+                () => codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken));
+
+            await Task.WhenAll(codeFixCollectionsTask, codeRefactoringsTask).ConfigureAwait(false);
+            return (await codeFixCollectionsTask.ConfigureAwait(false), await codeRefactoringsTask.ConfigureAwait(false));
```

Comment:
I discussed this with Cyrus recently and was told that it would probably be good to keep the WhenAll in here. I think the logic was that if both codeFixCollectionsTask and codeRefactoringsTask error or cancel, then using WhenAll allows us to retrieve both messages, while if we awaited them separately then we might only get one.

Reply:
Allison is correct.  It's an extremely subtle thing and easy to make a mistake around.  However, we can lose exceptions if we kick off tasks and then await each one serially.  then WhenAll ensures we properly see all of this, and doesn't hurt anything in terms of perf.  i.e. the later awaits will be no-ops because they will precheck if the tasks are complete and will grab the values if so.

Reply:
Ah interesting, that behavior makes sense now that I think about it.  Thanks for the explanation!